server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionWorker.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.iotdb.db.engine.compaction.task.CompactionTaskSummary;
+import org.apache.iotdb.db.service.metrics.recorder.CompactionMetricsRecorder;
+import org.apache.iotdb.db.utils.datastructure.FixedPriorityBlockingQueue;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class CompactionWorker implements Runnable {
+  private static final Logger log = LoggerFactory.getLogger("COMPACTION");
+  private final int threadId;
+  private final FixedPriorityBlockingQueue<AbstractCompactionTask> compactionTaskQueue;
+
+  public CompactionWorker(
+      int threadId, FixedPriorityBlockingQueue<AbstractCompactionTask> compactionTaskQueue) {
+    this.threadId = threadId;
+    this.compactionTaskQueue = compactionTaskQueue;
+  }
+
+  @Override
+  public void run() {
+    while (!Thread.currentThread().isInterrupted()) {
+      try {
+        AbstractCompactionTask task = compactionTaskQueue.take();
+        // add metrics
+        CompactionMetricsRecorder.recordTaskInfo(
+            task, CompactionTaskStatus.POLL_FROM_QUEUE, compactionTaskQueue.size());
+        if (task != null && task.checkValidAndSetMerging()) {
+          CompactionTaskSummary summary = task.getSummary();
+          CompactionTaskFuture future = new CompactionTaskFuture(summary);
+          CompactionTaskManager.getInstance().recordTask(task, future);
+          task.call();
+        }
+      } catch (InterruptedException e) {
+        log.error("CompactionThread-{} terminates because interruption", threadId);
+        return;
+      } catch (Throwable t) {
+        log.error("CompactionThread-{} meets exception", threadId, t);
+      }
+    }
+  }
+
+  class CompactionTaskFuture implements Future<CompactionTaskSummary> {
+    CompactionTaskSummary summary;
+
+    public CompactionTaskFuture(CompactionTaskSummary summary) {
+      this.summary = summary;
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+      summary.cancel();
+      return true;
+    }
+
+    @Override
+    public boolean isCancelled() {
+      return summary.isCancel();
+    }
+
+    @Override
+    public boolean isDone() {
+      return summary.isFinished();
+    }
+
+    @Override
+    public CompactionTaskSummary get() throws InterruptedException, ExecutionException {
+      while (!summary.isFinished()) {
+        TimeUnit.MILLISECONDS.sleep(100);
+      }
+      return summary;
+    }
+
+    @Override
+    public CompactionTaskSummary get(long timeout, @NotNull TimeUnit unit)
+        throws InterruptedException, ExecutionException, TimeoutException {
+      long perSleepTime = timeout > 100 ? timeout : 100;
+      long totalSleepTime = 0L;
+      while (!summary.isFinished()) {
+        if (totalSleepTime > timeout) {
```

Comment:
```suggestion
        if (totalSleepTime >= timeout) {
```

Or you will wait double time

Reply:
resolve